### DISCUSSION TITLE
error when using createArrowFiles

### DISCUSSION THREAD

**Discussion by danyama312** [GreenleafLab/ArchR#1417] [Questions / Documentation]
Hi all,

I am running the following code:

```
ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  minTSS = 4, #Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)
```
But get the following error message:

> 2022-05-03 14:50:29 : (Y_Naive_2 : 1 of 16) Successful creation of Temporary File, 6.742 mins elapsed.
> 2022-05-03 14:50:29 : (Y_Naive_2 : 1 of 16) Creating ArrowFile From Temporary File, 6.742 mins elapsed.
> 
> ************************************************************
> 2022-05-03 14:50:39 : ERROR Found in .tmpToArrow for (Y_Naive_2 : 1 of 16) 
> LogFile = ArchRLogs/ArchR-createArrows-288bc1b92a3dd-Date-2022-05-03_Time-14-43-45.log
> 
> <simpleError in h(simpleError(msg, call)): error in evaluating the argument 'i' in selecting a method for function '[': error in evaluating the argument 'x' in selecting a method for function 'which': no method found for function 'match' and signature XRawList, XRawList>
> 
> ************************************************************
> 
> 
> ************************************************************
> 2022-05-03 14:50:39 : ERROR Found in .tmpToArrow for (Y_Naive_2 : 1 of 16) 
> LogFile = ArchRLogs/ArchR-createArrows-288bc1b92a3dd-Date-2022-05-03_Time-14-43-45.log
> 
> <simpleError in .logError(e, fn = ".tmpToArrow", info = prefix, errorList = errorList,     logFile = logFile): Exiting See Error Above>
> 
> ************************************************************
> 
> 2022-05-03 14:50:39 : createArrowFiles has encountered an error, checking if any ArrowFiles completed..
> ArchR logging successful to : ArchRLogs/ArchR-createArrows-288bc1b92a3dd-Date-2022-05-03_Time-14-43-45.log

The structure of inputFiles (total of 16 fragment files) is as follows:

```
> head(inputFiles)
                                                           O_14D_1 
"/home/rstudio/cellRanger_Fragments/O_14D_1/outs/fragments.tsv.gz" 
                                                           O_14D_2 
"/home/rstudio/cellRanger_Fragments/O_14D_2/outs/fragments.tsv.gz" 
                                                           O_30D_1 
"/home/rstudio/cellRanger_Fragments/O_30D_1/outs/fragments.tsv.gz" 
                                                           O_30D_2 
"/home/rstudio/cellRanger_Fragments/O_30D_2/outs/fragments.tsv.gz" 
                                                            O_5D_1 
 "/home/rstudio/cellRanger_Fragments/O_5D_1/outs/fragments.tsv.gz" 
                                                            O_5D_2 
 "/home/rstudio/cellRanger_Fragments/O_5D_2/outs/fragments.tsv.gz" 
```

` sessionInfo()`

> R version 4.1.2 (2021-11-01)
> Platform: x86_64-pc-linux-gnu (64-bit)
> Running under: Ubuntu 20.04.3 LTS
> 
> Matrix products: default
> BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
> 
> locale:
>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
>  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
> [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
> 
> attached base packages:
>  [1] parallel  splines   grid      stats4    stats     graphics  grDevices utils     datasets  methods  
> [11] base     
> 
> other attached packages:
>  [1] Rsamtools_2.10.0                   BSgenome.Mmusculus.UCSC.mm10_1.4.3 BSgenome_1.62.0                   
>  [4] rtracklayer_1.54.0                 Biostrings_2.62.0                  XVector_0.34.0                    
>  [7] optparse_1.7.1                     scales_1.2.0                       RColorBrewer_1.1-3                
> [10] ggridges_0.5.3                     patchwork_1.1.1                    tictoc_1.0.1                      
> [13] reshape2_1.4.4                     ggpubr_0.4.0                       future.apply_1.9.0                
> [16] future_1.25.0                      ArchR_1.0.2                        magrittr_2.0.3                    
> [19] rhdf5_2.38.1                       data.table_1.14.2                  forcats_0.5.1                     
> [22] stringr_1.4.0                      dplyr_1.0.9                        purrr_0.3.4                       
> [25] readr_2.1.2                        tidyr_1.2.0                        tibble_3.1.6                      
> [28] tidyverse_1.3.1                    Signac_1.6.0                       Seurat_4.1.0                      
> [31] cicero_1.12.0                      monocle_2.22.0                     DDRTree_0.1.5                     
> [34] irlba_2.3.5                        VGAM_1.1-6                         ggplot2_3.3.5                     
> [37] Matrix_1.3-4                       Gviz_1.38.4                        SeuratObject_4.1.0                
> [40] monocle3_1.2.8                     SingleCellExperiment_1.16.0        SummarizedExperiment_1.24.0       
> [43] GenomicRanges_1.46.1               GenomeInfoDb_1.30.1                IRanges_2.28.0                    
> [46] S4Vectors_0.32.4                   MatrixGenerics_1.6.0               matrixStats_0.62.0                
> [49] Biobase_2.54.0                     BiocGenerics_0.40.0               
> 
> loaded via a namespace (and not attached):
>   [1] Hmisc_4.7-0              ica_1.0-2                RcppRoll_0.3.0           ps_1.7.0                
>   [5] lmtest_0.9-40            rprojroot_2.0.3          crayon_1.5.1             spatstat.core_2.4-2     
>   [9] MASS_7.3-54              rhdf5filters_1.6.0       nlme_3.1-153             backports_1.4.1         
>  [13] qlcMatrix_0.9.7          reprex_2.0.1             rlang_1.0.2              ROCR_1.0-11             
>  [17] readxl_1.4.0             nloptr_2.0.0             callr_3.7.0              limma_3.50.3            
>  [21] filelock_1.0.2           BiocParallel_1.28.3      rjson_0.2.21             bit64_4.0.5             
>  [25] glue_1.6.2               pheatmap_1.0.12          sctransform_0.3.3        processx_3.5.3          
>  [29] spatstat.sparse_2.1-1    AnnotationDbi_1.56.2     spatstat.geom_2.4-0      haven_2.5.0             
>  [33] tidyselect_1.1.2         usethis_2.1.5            fitdistrplus_1.1-8       XML_3.99-0.9            
>  [37] zoo_1.8-10               GenomicAlignments_1.30.0 xtable_1.8-4             cli_3.3.0               
>  [41] zlibbioc_1.40.0          rstudioapi_0.13          miniUI_0.1.1.1           rpart_4.1-15            
>  [45] fastmatch_1.1-3          ensembldb_2.18.4         shiny_1.7.1              xfun_0.30               
>  [49] pkgbuild_1.3.1           cluster_2.1.2            KEGGREST_1.34.0          ggrepel_0.9.1           
>  [53] biovizBase_1.42.0        listenv_0.8.0            png_0.1-7                withr_2.5.0             
>  [57] lsa_0.73.2               bitops_1.0-7             slam_0.1-50              ggforce_0.3.3           
>  [61] plyr_1.8.7               cellranger_1.1.0         sparsesvd_0.2            AnnotationFilter_1.18.0 
>  [65] pillar_1.7.0             cachem_1.0.6             GenomicFeatures_1.46.5   fs_1.5.2                
>  [69] vctrs_0.4.1              ellipsis_0.3.2           generics_0.1.2           devtools_2.4.3          
>  [73] tools_4.1.2              foreign_0.8-81           munsell_0.5.0            tweenr_1.0.2            
>  [77] DelayedArray_0.20.0      fastmap_1.1.0            HSMMSingleCell_1.14.0    compiler_4.1.2          
>  [81] pkgload_1.2.4            abind_1.4-5              httpuv_1.6.5             sessioninfo_1.2.2       
>  [85] plotly_4.10.0            GenomeInfoDbData_1.2.7   gridExtra_2.3            lattice_0.20-45         
>  [89] deldir_1.0-6             utf8_1.2.2               later_1.3.0              BiocFileCache_2.2.1     
>  [93] jsonlite_1.8.0           docopt_0.7.1             pbapply_1.5-0            carData_3.0-5           
>  [97] lazyeval_0.2.2           promises_1.2.0.1         car_3.0-12               latticeExtra_0.6-29     
> [101] goftest_1.2-3            spatstat.utils_2.3-0     reticulate_1.24          checkmate_2.1.0         
> [105] cowplot_1.1.1            Rtsne_0.16               dichromat_2.0-0          uwot_0.1.11             
> [109] igraph_1.3.1             survival_3.2-13          yaml_2.3.5               htmltools_0.5.2         
> [113] memoise_2.0.1            VariantAnnotation_1.40.0 BiocIO_1.4.0             viridisLite_0.4.0       
> [117] digest_0.6.29            assertthat_0.2.1         mime_0.12                rappdirs_0.3.3          
> [121] densityClust_0.3.2       RSQLite_2.2.13           remotes_2.4.2            blob_1.2.3              
> [125] fastICA_1.2-3            Formula_1.2-4            Rhdf5lib_1.16.0          Cairo_1.5-15            
> [129] ProtGenerics_1.26.0      RCurl_1.98-1.6           broom_0.8.0              hms_1.1.1               
> [133] modelr_0.1.8             colorspace_2.0-3         base64enc_0.1-3          BiocManager_1.30.17     
> [137] nnet_7.3-16              Rcpp_1.0.8.3             RANN_2.6.1               ggseqlogo_0.1           
> [141] fansi_1.0.3              tzdb_0.3.0               brio_1.1.3               parallelly_1.31.1       
> [145] SnowballC_0.7.0          R6_2.5.1                 lifecycle_1.0.1          curl_4.3.2              
> [149] ggsignif_0.6.3           minqa_1.2.4              leiden_0.3.10            testthat_3.1.4          
> [153] getopt_1.20.3            desc_1.4.1               RcppAnnoy_0.0.19         htmlwidgets_1.5.4       
> [157] polyclip_1.10-0          biomaRt_2.50.3           terra_1.5-21             rvest_1.0.2             
> [161] mgcv_1.8-38              globals_0.14.0           htmlTable_2.4.0          spatstat.random_2.2-0   
> [165] codetools_0.2-18         lubridate_1.8.0          FNN_1.1.3                prettyunits_1.1.1       
> [169] dbplyr_2.1.1             gtable_0.3.0             DBI_1.1.2                tensor_1.5              
> [173] httr_1.4.2               KernSmooth_2.23-20       stringi_1.7.6            progress_1.2.2          
> [177] farver_2.1.0             viridis_0.6.2            xml2_1.3.3               combinat_0.0-8          
> [181] boot_1.3-28              lme4_1.1-29              restfulr_0.0.13          scattermore_0.8         
> [185] bit_4.0.4                jpeg_0.1-9               spatstat.data_2.2-0      pkgconfig_2.0.3         
> [189] rstatix_0.7.0            knitr_1.39

Note that I have tried Archr 1.0.2 and 1.0.1, but get the same error message as above. I have tried both addArchRThreads(threads = 1) 
and addArchRThreads(threads = 16) as suggested elsewhere but this has not worked either. 

Any suggestions would be appreciated!

**Comment by danyama312** [comment]
Hi all,

I have also loaded the libraries "parallel" and "magick" as other threads have suggested for those using R 4.1.2, but still run into the same issue. I have attempted using the tutorial data also, but this has not been successful.

```
inputFiles <- getTutorialData("Hematopoiesis")
inputFiles


ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  minTSS = 4, #Dont set this too high because you can always increase later
  minFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)
ArrowFiles
```

Please find the log file below, apologies for not attaching it earlier!
[ArchR-createArrows-2a311776702f3-Date-2022-05-11_Time-15-08-30.log](https://github.com/GreenleafLab/ArchR/files/8671164/ArchR-createArrows-2a311776702f3-Date-2022-05-11_Time-15-08-30.log)

**Reply by rcorces** [reply]
thanks for uploading your logfile. Note that errors and bug reports belong in Issues whereas usage questions and feature requests belong in Discussions.

For your error, it could be something related to `data.table`.
Try loading a clean R session and then running
```
remove.packages("data.table")
install.packages("data.table", type = "source", repos = "https://Rdatatable.gitlab.io/data.table")
```

Then retry with the tutorial data and code from scratch

**Reply by danyama312** [reply]
Thank you very much for your prompt response.

Unfortunately the code above did not solve the problem.
Please find my log file below if that would help.
[ArchR-createArrows-2bca714a0026a-Date-2022-05-11_Time-15-35-34.log](https://github.com/GreenleafLab/ArchR/files/8671388/ArchR-createArrows-2bca714a0026a-Date-2022-05-11_Time-15-35-34.log)

**Reply by rcorces** [reply]
I dont have any additional suggestions, sorry. If the tutorial code doesnt work for you, its something with your environment and that is very challenging to figure out.

**Reply by danyama312** [reply]
Just thought I would note that I eventually got it to work by using a fresh environment (i.e installing everything from scratch) on R 4.1.0.

**Reply by rcorces** [reply]
thanks for reporting back. I'm glad that this is now working for you and that is helpful information to know. I wish we knew what caused these sporadic environment-specific errors.

**Comment by liuyifang** [comment]
I also meet this issue. I am using macOS M1 chip. By remove and reinstall R following https://stackoverflow.com/questions/70638118/configuring-compilers-on-mac-m1-big-sur-monterey-for-rcpp-and-other-tools
seems solved this problem.

**Comment by rcorces** [comment]
Thanks for reminding me about this. For what it is worth, I also encountered this relatively cryptic error recently and I think I fixed it on the `dev` branch via https://github.com/GreenleafLab/ArchR/pull/1795/commits/b1f5c07b79f42130eea23a9e32e7ee1f6b298782
I'm not sure why the error crops up or why it seems to be sporadic but in my hands, this commit fixed the problem without needed to reinstall anything.
